From 8f54034221a42d1677ada3b921fc533195f3b393 Mon Sep 17 00:00:00 2001
From: renchenglei <chenglei.ren@intel.com>
Date: Fri, 2 Apr 2021 18:03:24 +0800
Subject: [PATCH 02/14] INTERNAL: Don't close LG client when spice is
 disconneted

We need support multi LG client, we need keep LG client
alive when spice is used by other client.
---
 client/src/main.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/client/src/main.c b/client/src/main.c
index 9ce9384..6694c2d 100644
--- a/client/src/main.c
+++ b/client/src/main.c
@@ -43,6 +43,8 @@ Place, Suite 330, Boston, MA 02111-1307 USA
 #include "kb.h"
 #include "ll.h"
 
+static bool spice_running = true;
+
 // forwards
 static int cursorThread(void * unused);
 static int renderThread(void * unused);
@@ -386,12 +388,13 @@ static int frameThread(void * unused)
 int spiceThread(void * arg)
 {
   while(state.running)
-    if (!spice_process())
+    if (spice_running && !spice_process())
     {
       if (state.running)
       {
-        state.running = false;
         DEBUG_ERROR("failed to process spice messages");
+        spice_running = false;
+        continue;
       }
       break;
     }
@@ -1313,6 +1316,19 @@ int run()
           closeAlert  = NULL;
         }
       }
+      if (!spice_running && params.useSpiceInput && spice_ready())
+      {
+        for(int i = 0; i < SDL_NUM_SCANCODES; ++i)
+          if (state.keyDown[i])
+          {
+            uint32_t scancode = mapScancode(i);
+            if (scancode == 0)
+              continue;
+            state.keyDown[i] = false;
+            spice_key_up(scancode);
+          }
+        spice_disconnect();
+      }
     }
 
     break;
-- 
2.25.1

